pennylane/devices/qubit/sampling.py

```diff
@@ -65,6 +65,8 @@ def _group_measurements(mps: List[Union[SampleMeasurement, ClassicalShadowMP, Sh
     mp_no_obs_indices = []
 
     for i, mp in enumerate(mps):
+        if isinstance(mp.obs, Sum):
+            mps[i].obs = qml.simplify(mp.obs)
```

Comment:
This still give us some issues if the Sum is nested, like this example:

```
dev = qml.device("default.qubit", wires=2, shots=1000)

@qml.qnode(dev)
def circuit():
    qml.RX(np.pi/2, 0)
    qml.RX(-np.pi/2, 1)
    return qml.expval(qml.Y(1)), qml.expval(2 * (qml.Y(0)+qml.Y(0)+qml.Y(0)))

split_circuit = qml.transforms.split_non_commuting(circuit)

circuit(), split_circuit()
```
Might be enough to simplify `SProd` before checking for a sum, I can't think of another case where this will happen, but maybe I'm just not being creative enough

Reply:
Good point. `SProd` currently does not flatten into multiple terms when the base is `Sum`, so we need to simplify scalar products as well here

Reply:
https://github.com/PennyLaneAI/pennylane/pull/5885

This PR will change this

Reply:
Why does `Sum` require simplification, but not `Prod`?

Reply:
>  I imagine we could have a Prod where one of the operands is a Sum with duplicate terms? What would happen in this case?

This works fine (I had already tried)

```
import numpy as np
import pennylane as qml

dev = qml.device("default.qubit", wires=2, shots=1000)


@qml.qnode(dev)
def circuit():
    qml.RX(np.pi / 2, 0)
    qml.RX(-np.pi / 2, 1)
    return (
        qml.expval(qml.Y(0) + qml.Y(0) + qml.Y(0)),
        qml.expval((qml.Y(0) + qml.Y(0) + qml.Y(0)) @ (qml.Y(0) + qml.Y(0) + qml.Y(0))),
        qml.expval(qml.Y(0) @ (qml.Y(0) + qml.Y(0) + qml.Y(0))),
        qml.expval(2 * qml.Y(0) @ (qml.Y(0) + qml.Y(0) + qml.Y(0))),
    )


transf_circ = qml.transforms.split_non_commuting(circuit)

assert (transf_circ(), circuit())
```

Reply:
Still, there might be cases not considered here where also `Prod` must be simplified, so I think simplifying `Prod` as well is a good suggestion. I just pushed another commit 👍